woocommerce/woocommerce-ios
# Edit Product > Price Settings: extract view data logic from view controller

Part of #1985 

Apology for a bigger PR than I would hope 🙇🏻‍♀️more than half of the changes are in the unit tests for the price settings actions, which we couldn't have before. Please feel free to suggest any other naming for the new classes! I picked `*ViewModel` for the main class just because we've had a few view models in the codebase.

## Changes

- Created 2 protocols:
  - `ProductPriceSettingsViewModelOutput` that contains the properties needed to configure the price settings UI
  - `ProductPriceSettingsActionHandler` that handles actions related to the price settings data
- Created `ProductPriceSettingsViewModel` that implements the above 2 protocols with `ProductPriceSetingsError` error enum that could occur when saving price settings
- Moved view data logic from `ProductPriceSettingsViewController` to `ProductPriceSettingsViewModel`
- Replaced `product` param with `siteID` in `ProductTaxClassListSelectorDataSource` init, so that its caller `ProductPriceSettingsViewController` doesn't have access to the `product` and updated its unit tests
- New in unit tests:
  - Added support for price settings fields in `MockProduct`
  - Created `MockTaxClassStoresManager` to mock the action handler for `TaxClassAction` for retrieving a tax class
  - Added unit tests for all functions in `ProductPriceSettingsActionHandler`, including a few commented out cases that are pending separate issues (discovered by the unit tests)

## Testing

There should be no changes to the existing price settings UX. I'd recommend the following sanity check:

- Go to the Products tab
- Tap on a simple Product
- Tap on the price row
- Edit a few fields of your choice, maybe starting with simple changes because we are going to revert soon
- Tap "<" in the navigation bar --> there should be a discard changes action sheet
- Dismiss the action sheet
- Revert the changes you made in step 4
- Tap "<" in the navigation bar --> there should **not** be a discard changes action sheet and it goes back to the edit product main screen
- Tap on the price row
- Edit a few fields of your choice
- Tap "Done" --> it should go back to the edit product main screen with the new price settings in the price row
- Tap "Update" and wait for the API update to complete --> the product should have the updated price settings remotely and in the app

Update release notes:
